csharp/src/IceDiscovery/Lookup.cs

```diff
@@ -450,77 +370,146 @@ internal void AdapterRequestTimedOut(AdapterRequest request)
                 }
             }
         }
-
-        internal Timer Timer() => _timer;
-
-        internal int LatencyMultiplier() => _latencyMultiplier;
-
-        private readonly LocatorRegistry _registry;
-        private readonly ILookupPrx _lookup;
-        private readonly Dictionary<ILookupPrx, ILookupReplyPrx?> _lookups = new Dictionary<ILookupPrx, ILookupReplyPrx?>();
-        private readonly int _timeout;
-        private readonly int _retryCount;
-        private readonly int _latencyMultiplier;
-        private readonly string _domainId;
-
-        private readonly Timer _timer;
-        private bool _warnOnce = true;
-        private readonly Dictionary<Identity, ObjectRequest> _objectRequests = new Dictionary<Identity, ObjectRequest>();
-        private readonly Dictionary<string, AdapterRequest> _adapterRequests = new Dictionary<string, AdapterRequest>();
     }
 
     internal class LookupReply : ILookupReply
     {
+        private readonly Lookup _lookup;
+
         public LookupReply(Lookup lookup) => _lookup = lookup;
 
-        public void FoundObjectById(Identity id, IObjectPrx? proxy, Current c)
-            => _lookup.FoundObject(id, c.Identity.Name, proxy!); // proxy cannot be null
+        public void FoundObjectById(Identity id, IObjectPrx? proxy, Current c) =>
+            _lookup.FoundObject(id, c.Identity.Name, proxy!); // proxy cannot be null
 
         public void FoundAdapterById(string adapterId, IObjectPrx? proxy, bool isReplicaGroup, Current c) =>
             _lookup.FoundAdapter(adapterId, c.Identity.Name, proxy!, isReplicaGroup); // proxy cannot be null
+    }
 
-        private readonly Lookup _lookup;
+    internal class ObjectRequest : Request<Identity>, ITimerTask
+    {
+        public void Response(IObjectPrx proxy) => Finished(proxy);
+
+        public override void Finished(IObjectPrx? proxy)
+        {
+            foreach (TaskCompletionSource<IObjectPrx?> cb in Callbacks)
+            {
+                cb.SetResult(proxy);
+            }
+            Callbacks.Clear();
+        }
+
+        public void RunTimerTask() => Lookup.ObjectRequestTimedOut(this);
+
+        protected override async Task InvokeWithLookup(string domainId, ILookupPrx lookup, ILookupReplyPrx lookupReply)
+        {
+            try
+            {
+                await lookup.FindObjectByIdAsync(domainId, Id, lookupReply).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Lookup.ObjectRequestException(this, ex);
+            }
+        }
+
+        internal ObjectRequest(Lookup lookup, Identity id, int retryCount)
+            : base(lookup, id, retryCount)
+        {
+        }
+
+    }
+
+    internal abstract class Request<T>
+    {
+        public T Id { get; }
+        public string RequestId { get; }
+
+        protected List<TaskCompletionSource<IObjectPrx?>> Callbacks = new List<TaskCompletionSource<IObjectPrx?>>();
```

Comment:
We could rename Callbacks to Waiters or Sources.

Reply:
Fixed